src/sentry/integrations/bitbucket_server/webhook.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import absolute_import
+
+import logging
+
+import six
+
+from datetime import datetime
+from django.db import IntegrityError, transaction
+from django.http import HttpResponse, Http404
+from django.utils import timezone
+from django.utils.decorators import method_decorator
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic.base import View
+from simplejson import JSONDecodeError
+from sentry.models import Commit, CommitAuthor, Organization, Repository
+from sentry.plugins.providers import IntegrationRepositoryProvider
+from sentry.utils import json
+
+logger = logging.getLogger("sentry.webhooks")
+
+PROVIDER_NAME = "integrations:bitbucket_server"
+
+
+class Webhook(object):
+    def __call__(self, organization, integration_id, event):
+        raise NotImplementedError
+
+    def update_repo_data(self, repo, event):
+        """
+        Given a webhook payload, update stored repo data if needed.
+        """
+
+        name_from_event = event["repository"]["project"]["key"] + "/" + event["repository"]["slug"]
+        # build the URL manually since it doesn't come back from the API in
+        # the form that we need
+        url_from_event = event["repository"]["links"]["self"][0]["href"]
+
+        if (
+            repo.name != name_from_event
+            or repo.config.get("name") != name_from_event
+            or repo.url != url_from_event
+        ):
+            repo.update(name=name_from_event, config=dict(repo.config, name=name_from_event))
+
+
+class PushEventWebhook(Webhook):
+    def __call__(self, organization, integration_id, event):
+        authors = {}
+
+        try:
+            repo = Repository.objects.get(
+                organization_id=organization.id,
+                provider=PROVIDER_NAME,
+                external_id=six.text_type(event["repository"]["id"]),
+            )
+        except Repository.DoesNotExist:
+            raise Http404()
+
+        client = repo.get_provider().get_installation(integration_id, organization.id).get_client()
+
+        # while we're here, make sure repo data is up to date
+        self.update_repo_data(repo, event)
+
+        parts = repo.name.split("/")
```

Comment:
@pierredup Related to my earlier comment about clarity, could we do something like this instead:
```
[project_name, repo_name] = repo.name.split("/")
```
That makes the subsequent invocation of `get_commits` a bit easier to understand.